libindex/fetcher.go

```diff
@@ -0,0 +1,380 @@
+		// Catch the old docker media type.
+		fallthrough
+	case ct == "application/gzip" || ct == "application/x-gzip":
+		// GHCR reports gzipped layers as the latter.
+		fallthrough
+	case strings.HasSuffix(ct, ".tar+gzip"):
+		g, err := gzip.NewReader(br)
+		if err != nil {
+			return "", err
+		}
+		defer g.Close()
+		r = g
+	case ct == "application/zstd":
+		fallthrough
+	case strings.HasSuffix(ct, ".tar+zstd"):
+		s, err := zstd.NewReader(br)
+		if err != nil {
+			return "", err
+		}
+		defer s.Close()
+		r = s
+	case ct == "application/x-tar":
+		fallthrough
+	case strings.HasSuffix(ct, ".tar"):
+		r = br
+	default:
+		return "", fmt.Errorf("fetcher: unknown content-type %q", ct)
+	}
+
+	buf := bufio.NewWriter(fd)
+	n, err := io.Copy(buf, r)
+	zlog.Debug(ctx).Int64("size", n).Msg("wrote file")
+	if err != nil {
+		return "", err
+	}
+	if err := buf.Flush(); err != nil {
+		return "", err
+	}
+	if got := vh.Sum(nil); !bytes.Equal(got, want) {
+		err := fmt.Errorf("fetcher: validation failed: got %q, expected %q",
+			hex.EncodeToString(got),
+			hex.EncodeToString(want))
+		return "", err
+	}
+
+	zlog.Debug(ctx).Msg("layer fetch ok")
+	rm = false
+	return name, nil
+}
+
+// Fetcher returns an indexer.Fetcher.
+func (f *FetchArena) Fetcher() *FetchProxy {
+	return &FetchProxy{f: f}
+}
+
+// FetchProxy tracks the files fetched for layers.
+//
+// This can be unexported if FetchArena gets unexported.
+type FetchProxy struct {
+	f     *FetchArena
+	mu    sync.Mutex
+	clean []string
+}
+
+// Fetch populates all the layers locally.
+func (f *FetchProxy) Fetch(ctx context.Context, ls []*claircore.Layer) error {
+	g, ctx := errgroup.WithContext(ctx)
+	for _, l := range ls {
+		g.Go(f.fetchOne(ctx, l))
+	}
+	if err := g.Wait(); err != nil {
+		return fmt.Errorf("encountered error while fetching a layer: %v", err)
+	}
+	return nil
+}
+
+// FetchOne runs a fetch though the singleflight while waiting on the passed-in
+// context.
+func (f *FetchProxy) fetchOne(ctx context.Context, l *claircore.Layer) func() error {
+	fn := func() (interface{}, error) {
+		return f.f.realizeLayer(ctx, l)
+	}
+	return func() error {
+		h := l.Hash.String()
+		select {
+		case res := <-f.f.sf.DoChan(h, fn):
+			if err := res.Err; err != nil {
+				return err
+			}
+			fn := res.Val.(string)
+			if err := l.SetLocal(fn); err != nil {
+				return err
+			}
+			if err := f.addRef(h); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+		return nil
+	}
+}
+
+func (f *FetchProxy) addRef(digest string) error {
+	if err := f.f.incRef(digest); err != nil {
+		return err
+	}
+	f.mu.Lock()
+	f.clean = append(f.clean, digest)
+	f.mu.Unlock()
+	return nil
+}
+
+// Close marks all the layers' backing files as unused.
+//
+// This method may actually delete the backing files.
+func (f *FetchProxy) Close() error {
+	var err error
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	for _, digest := range f.clean {
+		_, e := f.f.decRef(digest)
```

Comment:
Can we consider `fp` (and possibly `fa`) so lines like this (`fp.fa.decRef`) are a little easier to read?